### PR TITLE
Bug 1831632 - Update group membership report to more prominently display users are in groups that not Mozilla email addresses (i.e. personal accounts)

### PR DESCRIPTION
**Pull request by dklawren**
* Some cleanup of the group member report by adding a table for the different query fields
* Extend Bugzilla::Group->members_complete to take an argument that will filter results based on search type
* non-mozilla search type filters the users to only show email addresses that do not match the groups regex if one exists.
* The regex is used by Mozilla to automatically place users in specific confidential groups.
* Figured this would be better than hard-coding '.*@mozilla(foundation)?\.(org|com)' and all of the other possibilities.

### CODE REVIEW

File: Bugzilla/Group.pm [hunk 1]
```diff
@@ -119,13 +119,36 @@ sub members_complete {
...
+  # If we want a list of users that are members but do not match the regexp
+  # if the current group, then we need to filter the user list.
+  # Regexp are normally used to automatically add users to a group such
+  # as mozilla-employee-confidential, etc.
+  if ($type eq 'non-mozilla' && $self->user_regexp) {
+    my $regexp = $self->user_regexp;
+    foreach my $type (keys %{$result}) {
+      my @filtered;
+      foreach my $user (@{$result->{$type}}) {
+        push @filtered, $user if $user->login !~ m/$regexp/i;
```
Comment (cgsheeh) [inline Bugzilla/Group.pm:146]: Crafting a regex from user-input is a tad dangerous. In theory someone could create a very special regex that consumes a ton of memory, causing a DoS.

Is this report something that runs only for trusted users? Can we limit them to a selection, or maybe say "search for this string in the email" or something along those lines instead?

File: extensions/BMO/lib/Reports/Groups.pm [hunk 2]
```diff
@@ -185,18 +185,22 @@ sub members_report {
...
   my @group_names
     = sort grep { !/^(?:bz_.+|canconfirm|editbugs|editbugs-team|everyone)$/ }
     @grouplist;
   unshift(@group_names, '');
   $vars->{'groups'} = \@group_names;
 
+  my $include_disabled = $cgi->param('include_disabled') ? 1 : 0;
+  $vars->{'include_disabled'} = $include_disabled;
+
+  my $search_type = $cgi->param('search_type') || 'all';
```
Comment (cgsheeh) [inline extensions/BMO/lib/Reports/Groups.pm:198]: Can we limit `search_type` to only a set of allowed values here? Or perhaps in the `members_complete` subroutine.

**Review by cgsheeh — CHANGES_REQUESTED** [review]

File: Bugzilla/Group.pm (hunk 1 above)
Comment (dklawren) [reply Bugzilla/Group.pm:146]: The `$self->user_regexp` value is not coming from user input but rather from the Group object and is only edited by a group admin. For example the group `mozilla-corporation` has a $self->user_regexp value of `@mozilla\.com$ ` so the above line would be the same as `dkl@mozilla !~ m/@mozilla\.com$/`

The user_regexp value requires a user with 'editgroups` permissions to update the value.

File: extensions/BMO/lib/Reports/Groups.pm (hunk 2 above)
Comment (dklawren) [reply extensions/BMO/lib/Reports/Groups.pm:198]: Sure. Will do that.

File: Bugzilla/Group.pm [hunk 3]
```diff
@@ -119,13 +119,36 @@ sub members_complete {
...
+
   foreach my $group_id (@{$self->flatten_group_membership($self->id)}) {
     next if $group_id == $self->id;
     my $group_name = Bugzilla::Group->new({id => $group_id, cache => 1})->name;
     my $user_ids = $dbh->selectcol_arrayref($sth, undef, $group_id);
     $result->{$group_name} = Bugzilla::User->new_from_list($user_ids);
   }
+
+  # If we want a list of users that are members but do not match the regexp
+  # if the current group, then we need to filter the user list.
```
Comment (cgsheeh) [inline Bugzilla/Group.pm:138]: I think this should say "of the current group"?

**Review by cgsheeh — APPROVED** [review]

**Pull request merged**